mesonbuild/interpreter.py

```diff
@@ -2792,34 +2795,66 @@ def func_add_languages(self, node, args, kwargs):
             return False
         return self.add_languages(args, required)
 
-    def get_message_string_arg(self, node):
-        # reduce arguments again to avoid flattening posargs
-        (posargs, _) = self.reduce_arguments(node.args)
-        if len(posargs) != 1:
-            raise InvalidArguments('Expected 1 argument, got %d' % len(posargs))
-
-        arg = posargs[0]
-        if isinstance(arg, list):
-            argstr = stringifyUserArguments(arg)
-        elif isinstance(arg, dict):
-            argstr = stringifyUserArguments(arg)
-        elif isinstance(arg, str):
-            argstr = arg
-        elif isinstance(arg, int):
-            argstr = str(arg)
-        else:
-            raise InvalidArguments('Function accepts only strings, integers, lists and lists thereof.')
-
-        return argstr
+    def get_message_string_arg(self, args):
+        if len(args) != 1:
+            raise InvalidArguments('Expected 1 argument, got %d' % len(args))
+        return stringifyUserArguments(args[0])
 
+    @noArgsFlattening
     @noKwargs
     def func_message(self, node, args, kwargs):
-        argstr = self.get_message_string_arg(node)
+        argstr = self.get_message_string_arg(args)
         self.message_impl(argstr)
 
     def message_impl(self, argstr):
         mlog.log(mlog.bold('Message:'), argstr)
 
+    @noArgsFlattening
+    @FeatureNew('summary', '0.53.0')
+    def func_summary(self, node, args, kwargs):
+        sections = self.summary.setdefault(self.subproject, [])
+        if len(args) != 2:
+            raise InterpreterException('Summary accepts exactly two arguments.')
+        name, value = args
+        if not isinstance(name, str):
+            raise InterpreterException('Argument 1 must be a string.')
+        if name in [section[0] for section in sections]:
```

Comment:
this looks wrong, for one `in` is the wrong thing to use because a substring would match, plus as written we'd always iterate all of the sections, I think
```python
if any(name == s[0] for s in sections):
```

Reply:
Although, actually now that I think about it, maybe it would be better to use a `collections.OderedDict` instead of a list of tuples since then we could just do `if name in sections`?

Reply:
All dictionaries are ordered in python now, but I guess it's better make it explicit that we rely on ordering.